docs/changes/cosmology/11778.api.rst

```diff
@@ -0,0 +1,5 @@
+Rename ``z_at_value`` to ``z_at_scalar_value`` and add a wrapper function
```

Comment:
The overhead from setting up the iterator is likely small compared to the actual function call, so might it make sense to just state that `z_at_value` can now handle arrays? I.e., not mention `z_at_scalar_value` at all?

Reply:
> The overhead from setting up the iterator is likely small compared to the actual function call, 

That was the one clocked in https://github.com/astropy/astropy/pull/11778#issuecomment-848181639, right?

Reply:
That was a previous implementation. I'll reclock.

Reply:
It's a pretty negligible difference.

<img width="983" alt="Screen Shot 2021-07-16 at 12 16 45" src="https://user-images.githubusercontent.com/8949649/125978077-11b679ae-5cfd-42d7-b8b8-bdd3482b054f.png">

Reply:
The extra overhead from calling `z_at_value` on an array is expected? Not a serious difference, just wondering where that additional time  comes in...

Reply:
Looking at the above, I'm inclined to agree.
The point of ``z_at_scalar_value`` public  is that it's more performant. Except it basically isn't. I'm going to make it a private function. The interpolated case (#11361) should introduce a new function, perhaps ``z_at_interp_value``. If I can make ``z_at_scalar_value`` more performant relative to ``z_at_value``, I'll make it public.

Reply:
@dhomeier - list comprehension is really quite fast, hard to `np.nditer` to match, with all the broadcasting etc.

Reply:
> The point of `z_at_scalar_value` public is that it's more performant. Except it basically isn't. I'm going to

Well, 12-15 % faster is still faster, even if perhaps not enough to justify keeping it private. What was more surprising to me was that even looping over it 100 times is still 15 % faster than calling `z_at_value` directly on the value. Unless you are seeing some caching artefacts in `timeit`, but I can't really see where they'd come in.

Reply:
> @dhomeier - list comprehension is really quite fast, hard to `np.nditer` to match, with all the broadcasting etc.

Sounds sensible – I had not looked at the internals since my early incomplete review, so I thought there was still some code that the vectorised version did not execute on every single element of `fval`; but seeing now that there is also a bunch of extras included for accepting sequences of `bracket` etc.

Reply:
That's probably from the fact that ``z_at_value`` needs to consider how to broadcast all the values. I think if I were to do array nesting and so forth we would see that the manual implementation is equivalent to ``z_at_value``. ``z_at_value`` just doesn't know when it can simplify. If python allowed for multiple dispatch..